src/core/libmaven/isotopeDetection.cpp

```diff
@@ -63,207 +63,182 @@ void IsotopeDetection::pullIsotopes(PeakGroup* parentgroup)
+            filterIsotopePeaks(parentPeak, allPeaks, sample);
 
-            delete(eic);
             // find nearest peak as long as it is within RT window
-            float maxRtDiff=_mavenParameters->maxIsotopeScanDiff * _mavenParameters->avgScanTime;
-            //why are we even doing this calculation, why not have the parameter be in units of RT?
-            Peak* nearestPeak = NULL;
-            float d = FLT_MAX;
-            for (unsigned int i = 0; i < allPeaks.size(); i++) {
-                Peak& x = allPeaks[i];
-                float dist = abs(x.rt - rt);
-                if (dist > maxRtDiff)
+            Peak* nearestPeak = nullptr;
+            int dist = numeric_limits<int>::max();
+            for (auto& peak : allPeaks) {
+                int d = 0;
+                int minScan = min(peak.scan, parentPeak->scan);
+                int maxScan = max(peak.scan, parentPeak->scan);
+                while (minScan != maxScan) {
+                    auto scanInBetween = sample->getScan(++minScan);
+                    if (scanInBetween->mslevel == 1)
+                        ++d;
+                }
+                if (d > _mavenParameters->maxIsotopeScanDiff)
                     continue;
-                if (dist < d) {
-                    d = dist;
-                    nearestPeak = &x;
+                if (d < dist) {
+                    dist = d;
+                    nearestPeak = &peak;
                 }
             }
 
-            //delete (nearestPeak);
-            if (nearestPeak) { //if nearest peak is present
-                if (isotopes.count(isotopeName) == 0) { //label the peak of isotope
-                    PeakGroup g;
-                    g.meanMz = isotopeMass; //This get's updated in groupStatistics function
-                    g.expectedMz = isotopeMass;
-                    g.tagString = isotopeName;
-                    g.expectedAbundance = expectedAbundance;
-                    g.isotopeC13count = x.C13;
-                    g.setSelectedSamples(parentgroup->samples);
-
-                    // create a slice for this group; RT will be updated later
-                    mzSlice childSlice(mzmin,
-                                       mzmax,
-                                       0.0f,
-                                       numeric_limits<float>::max());
-                    g.setSlice(childSlice);
-                    isotopes[isotopeName] = g;
-                }
-                isotopes[isotopeName].addPeak(*nearestPeak); //add nearestPeak to isotope peak list
+            if (nearestPeak == nullptr) {
+                delete eic;
+                continue;
+            }
+
+            if (_mavenParameters->linkIsotopeRtRange) {
+                eic->adjustPeakBounds(*nearestPeak,
+                                      parentPeak->rtmin,
+                                      parentPeak->rtmax);
+                eic->getPeakDetails(*nearestPeak);
```

Comment:
peakFiltering should happen here instead? since the baseline and area have both changed now.

Reply:
Right. But what if the nearest peak is filtered and there was another (second-nearest) peak that was worthy of being the isotope's peak but it was earlier filtered due to distance? Very unlikely, but possible. Instead, while finding nearest peak, we change each peak's boundaries, apply peak filter on it, and only when it passes will its distance be considered.

Reply:
that makes sense